### PR TITLE
Pattern api

### PR DESCRIPTION
**Pull request by llllllllll**
pr against master not code-object

### CODE REVIEW

File: codetransformer/code.py [hunk 1]
```diff
@@ -458,13 +464,23 @@ def is_coroutine(self):
...
     def is_iterable_coroutine(self):
-        """Is this an async iterator defined with __anext__?
+        """Is this an async generator defined with types.coroutine?
 
         This is 3.5 and greater.
         """
         return bool(self._flags & Flags.CO_ITERABLE_COROUTINE)
 
     @property
+    def has_new_locals(self):
```
Comment (ssanderson) [inline codetransformer/code.py]: `creates_new_locals`?

File: codetransformer/code.py [hunk 2]
```diff
@@ -534,5 +547,21 @@ def bytecode_offset(self, instr):
...
+    def __getitem__(self, key):
+        return self.instrs[key]
+
     def __iter__(self):
         return iter(self.instrs)
+
+    def __len__(self):
+        return len(self.instrs)
+
+    def __contains__(self, instr):
```
Comment (ssanderson) [inline codetransformer/code.py:562]: Can we not just dispatch to `tuple.__contains__`?

File: codetransformer/core.py [hunk 3]
```diff
@@ -16,15 +18,26 @@ def _a_if_not_none(a, b):
...
+    """Exception raised to indicate that the ``code` or ``startcode``
+    attribute was accessed outside of a code context.
     """
     def __init__(self):
-        return super().__init__('no code context')
+        return super().__init__('no context')
 
 
-class CodeTransformer(object):
+class CodeTransformerMeta(type):
```
Comment (ssanderson) [inline codetransformer/core.py:34]: Can we get a docstring explaining what this does?

File: codetransformer/core.py [hunk 4]
```diff
@@ -16,15 +18,29 @@ def _a_if_not_none(a, b):
...
+    """Meta class for CodeTransformer to collect all pf the patterns
```
Comment (ssanderson) [inline codetransformer/core.py]: of?

Comment (ssanderson) [inline codetransformer/core.py]: Can we document where these patterns actually come from (e.g. note that we're collection methods decorated in a certain way)?

File: codetransformer/core.py [hunk 5]
```diff
@@ -133,23 +130,45 @@ def visit(self, code, *, name=None, filename=None, lnotab=None):
...
 
         instrs, varnames = tuple(zip(*reversed_varnames.items())) or ((), ())
-        for instr, varname in zip(instrs, self.visit_varnames(varnames)):
+        for instr, varname in zip(instrs, self.transform_varnames(varnames)):
             instr.arg = varname
 
-        with self._new_code_context(code):
+        with self._new_context(code):
+            instrs = tuple(code)
+            compiled_instrs = bytes(map(attrgetter('opcode'), instrs))
```
Comment (ssanderson) [inline codetransformer/core.py]: opcodes?

File: codetransformer/core.py [hunk 6]
```diff
@@ -133,23 +130,45 @@ def visit(self, code, *, name=None, filename=None, lnotab=None):
...
+            len_instrs = len(instrs)
```
Comment (ssanderson) [inline codetransformer/core.py]: I don't think we use this value outside of the while loop

File: codetransformer/core.py [hunk 7]
```diff
@@ -133,23 +130,45 @@ def visit(self, code, *, name=None, filename=None, lnotab=None):
...
+            extend_new = post_transform.extend
+            dispatcher = self._patterndispatcher
+            while idx < len_instrs:
+                try:
+                    processed, nconsumed = dispatcher(
+                        compiled_instrs[idx:],
+                        instrs[idx:],
+                        self.startcode
+                    )
+                except KeyError:
```
Comment (ssanderson) [inline codetransformer/core.py]: I'd rather use a more specific error here.

File: codetransformer/core.py [hunk 8]
```diff
@@ -133,23 +130,45 @@ def visit(self, code, *, name=None, filename=None, lnotab=None):
...
+                    append_new(instrs[idx])
+                    idx += 1
+                else:
+                    extend_new(processed)
```
Comment (ssanderson) [inline codetransformer/core.py]: I'm not a fan of doing this if it's not for performance.

File: codetransformer/core.py [hunk 9]
```diff
@@ -176,12 +195,14 @@ def __call__(self, f, *,
         )
 
     @contextmanager
-    def _new_code_context(self, code):
+    def _new_context(self, code):
         self._code_stack.append(code)
+        self._startcode_stack.append(0)
```
Comment (ssanderson) [inline codetransformer/core.py]: I read this as a magic number.

File: codetransformer/core.py [hunk 10]
```diff
@@ -190,4 +211,26 @@ def code(self):
...
+    def begin(self, startcode):
+        """Begin a new startcode.
+
+        Parameters
+        ----------
+        startcode : any
+            The startcode to begin.
+        """
+        try:
+            self._startcode_stack[-1] = startcode
```
Comment (ssanderson) [inline codetransformer/core.py:242]: It's a little bit confusing that we're writing to the last value of a variable called `stack`.

File: codetransformer/instructions.py [hunk 11]
```diff
@@ -80,19 +76,26 @@ def __new__(mcls, name, bases, dict_, *, opcode=None):
...
-        dict_['reljmp'] = _immutableattr(opcode in hasjrel)
-        dict_['opname'] = _immutableattr(opname[opcode])
-        dict_['uses_name'] = _immutableattr(opname_ in _uses_name)
-        dict_['uses_varname'] = _immutableattr(opname_ in _uses_varname)
-        dict_['uses_free'] = _immutableattr(opname_ in _uses_free)
-        dict_['have_arg'] = _immutableattr(opcode >= HAVE_ARGUMENT)
+        dict_['opcode'] = immutableattr(opcode)
+        dict_['absjmp'] = immutableattr(opcode in hasjabs)
+        dict_['reljmp'] = immutableattr(opcode in hasjrel)
+        dict_['opname'] = immutableattr(opname[opcode])
```
Comment (ssanderson) [inline codetransformer/instructions.py]: You already did this above.

File: codetransformer/instructions.py [hunk 12]
```diff
@@ -80,19 +76,26 @@ def __new__(mcls, name, bases, dict_, *, opcode=None):
...
+        dict_['uses_name'] = immutableattr(opname_ in _uses_name)
+        dict_['uses_varname'] = immutableattr(opname_ in _uses_varname)
+        dict_['uses_free'] = immutableattr(opname_ in _uses_free)
+        dict_['have_arg'] = immutableattr(opcode >= HAVE_ARGUMENT)
         cls = mcls._type_cache[opcode] = super().__new__(
             mcls, opname[opcode], bases, dict_,
         )
         return cls
 
+    def mcompile(self):
```
Comment (ssanderson) [inline codetransformer/instructions.py:96]: to_regex?

File: codetransformer/patterns.py [hunk 13]
```diff
@@ -0,0 +1,289 @@
+from operator import methodcaller, index
+import re
+from types import MethodType
+
+from .utils.instance import instance
+from .utils.immutable import immutable
+
+
+mcompile = methodcaller('mcompile')
```
Comment (ssanderson) [inline codetransformer/patterns.py:10]: See above comment.

File: codetransformer/patterns.py [hunk 14]
```diff
@@ -0,0 +1,289 @@
...
+            patterns.append(self)
+        if isinstance(other, or_):
+            patterns.extend(other.matchables)
+        else:
+            patterns.append(other)
+
+        return or_(*patterns)
+
+    def __ror__(self, other):
+        # Flip the order on the or method
```
Comment (ssanderson) [inline codetransformer/patterns.py:53]: Should we check that `other` is matchable here?

File: codetransformer/patterns.py [hunk 15]
```diff
@@ -0,0 +1,289 @@
...
+    def __invert__(self):
+        return not_(self)
+
+    def __getitem__(self, key):
+        try:
+            n = index(key)
+        except TypeError:
+            pass
+        else:
+            return matchnm(self, n)
```
Comment (ssanderson) [inline codetransformer/patterns.py]: this is easy to misread as `matchname`.  match_range?  match_count?

File: codetransformer/patterns.py [hunk 16]
```diff
@@ -0,0 +1,289 @@
...
+    def mcompile(self):
+        return self.matchable.mcompile() + self.modifier.mcompile()
+
+    def __repr__(self):
+        return '%r[%r]' % (self.matchable, self.modifier)
+    __str__ = __repr__
+
+
+class meta:
+    """Class for meta patterns and pattern likes.
```
Comment (ssanderson) [inline codetransformer/patterns.py]: e.g. ...

File: codetransformer/patterns.py [hunk 17]
```diff
@@ -0,0 +1,289 @@
...
+        if isinstance(key, tuple) and len(key) in (1, 2):
+            return matchnm(self, *key)
+
+        if not isinstance(key, modifier):
+            raise TypeError('invalid modifier: {0}'.format(key))
+
+        return modified(self, key)
+
+
+class modified(immutable, matchable):
```
Comment (ssanderson) [inline codetransformer/patterns.py]: `postfix_modifier` maybe?  There are lots of logical modifiers here that aren't subclasses of this type.

File: codetransformer/patterns.py [hunk 18]
```diff
@@ -0,0 +1,289 @@
...
+            n = index(key)
+        except TypeError:
+            pass
+        else:
+            return matchnm(self, n)
+
+        if isinstance(key, tuple) and len(key) in (1, 2):
+            return matchnm(self, *key)
+
+        if not isinstance(key, modifier):
```
Comment (ssanderson) [inline codetransformer/patterns.py]: Slight preference for inverting the check here.

File: codetransformer/patterns.py [hunk 19]
```diff
@@ -0,0 +1,289 @@
...
+
+    def mcompile(self):
+        matchable = self.matchable
+        if isinstance(matchable, (seq, or_, not_)):
+            return b'((?!(' + matchable.mcompile() + b')).)*'
+
+        return b'[^' + matchable.mcompile() + b']'
+
+    def __repr__(self):
+        return '~' + _prepr(self.matchable)
```
Comment (ssanderson) [inline codetransformer/patterns.py:221]: The logic for prepr seems like it's tied to `not_`.  Should we just inline it here?

File: codetransformer/patterns.py [hunk 20]
```diff
@@ -0,0 +1,289 @@
...
+    __slots__ = '*matchables',
+
+    def mcompile(self):
+        return b'(' + b'|'.join(map(mcompile, self.matchables)) + b')'
+
+    def __repr__(self):
+        return ' | '.join(map(_prepr, self.matchables))
+
+
+class not_(immutable):
```
Comment (ssanderson) [inline codetransformer/patterns.py]: I think this should be matchable.

File: codetransformer/patterns.py [hunk 21]
```diff
@@ -0,0 +1,289 @@
...
+
+
+@instance
+class option(modifier):
+    """Modifier that matches zero or one of a pattern.
+    """
+    _token = b'?'
+
+
+class matchnm(immutable, meta, defaults={'m': None}):
```
Comment (ssanderson) [inline codetransformer/patterns.py]: Should this also be matchable?

File: codetransformer/patterns.py (hunk 19 above)
Comment (llllllllll) [inline codetransformer/patterns.py:221]: This is also used in seq and or_

File: codetransformer/instructions.py (hunk 12 above)
Comment (llllllllll) [inline codetransformer/instructions.py:96]: I like that we could change to some other backend without worrying. the fact that the compiled matches are regular expressions is only an implementation detail

File: codetransformer/patterns.py (hunk 21 above)
Comment (llllllllll) [inline codetransformer/patterns.py]: just made meta matchable.

File: codetransformer/code.py (hunk 2 above)
Comment (llllllllll) [inline codetransformer/code.py:562]: gh didn't mark as outdated, we do this now

File: codetransformer/core.py (hunk 3 above)
Comment (llllllllll) [inline codetransformer/core.py:34]: done

File: codetransformer/transformers/literals.py [hunk 22]
```diff
@@ -43,21 +49,58 @@ class overloaded_dicts(CodeTransformer):
...
+        # TOS1 = k
+        # TOS2 = v
+
+        yield instructions.ROT_TWO()
+        # TOS  = k
+        # TOS1 = __map__
+        # TOS2 = v
+
+        yield instructions.STORE_SUBSCR()
+        self.begin(IN_COMPREHENSION)
```
Comment (ssanderson) [inline codetransformer/transformers/literals.py:80]: Do we need the startcode anymore?

**Pull request merged**